src/compiler/checker.ts

```diff
@@ -13971,13 +13973,13 @@ namespace ts {
             });
         }
 
-        function getIndexedAccessTypeOrUndefined(objectType: Type, indexType: Type, accessNode?: ElementAccessExpression | IndexedAccessTypeNode | PropertyName | BindingName | SyntheticExpression, accessFlags = AccessFlags.None, aliasSymbol?: Symbol, aliasTypeArguments?: readonly Type[]): Type | undefined {
+        function getIndexedAccessTypeOrUndefined(objectType: Type, indexType: Type, noUncheckedIndexedAccessCandidate?: boolean, accessNode?: ElementAccessExpression | IndexedAccessTypeNode | PropertyName | BindingName | SyntheticExpression, accessFlags = AccessFlags.None, aliasSymbol?: Symbol, aliasTypeArguments?: readonly Type[]): Type | undefined {
             if (objectType === wildcardType || indexType === wildcardType) {
                 return wildcardType;
             }
 
-            const shouldIncludeUndefined =
-                compilerOptions.noUncheckedIndexedAccess &&
+            const shouldIncludeUndefined = noUncheckedIndexedAccessCandidate ||

```

Comment:
I find the presedence of this a bit confusing - can we parenthesize the `&&` subexpression? Or am I missing something? I think this the the only place an initial `true` value can be introduced for `noUncheckedIndexedAccessCandidate`, right?

Reply:
Yeah, it’s kind of weird, but here’s what’s happening: in the first pass of checking nodes in a source file, when we see an element access, we try to get its type with this function, passing `accessFlags` based on the syntax where we found it. If the `objectType` and `indexType` are non-generic, we can return a fully resolved type right away, and we can mix `undefined` into that if `accessFlags` said it was a reading operation in an expression position and the `indexType` matched an index signature on `objectType` (as opposed to a named property). But if the `objectType` or `indexType` is generic, all we can do is return an IndexedAccessType to be resolved later. The problem is that we don’t know _now_ whether the indexed access type will resolve to an index signature or a named property, and when we come back _later_ we won’t have the `accessFlags` that tell us whether this is a reading position. So we extract the relevant info from the `accessFlags` in the first pass and cache it on the IndexedAccessType. Later, when we have instantiations of the `objectType` and/or `indexType`, we actually call this _same function again_, except we’ve lost the `accessFlags`—but, standing in for them, we have the `noUncheckedIndexedAccessCandidate` property that we cached on the IndexedAccessType. So when this `noUncheckedIndexedAccessCandidate` parameter is true, it actually means “we don’t have the any/all the `accessFlags` anymore, but we previously extracted the relevant info from them, and they indicated that you should include `undefined`.”

This would probably be more clear if the first-pass checking and deferred checking bits were split into two separate function, but there are a good number of call sites so it would be kind of an unwieldy change.

Reply:
Right, right - I just think that for this condition writing `a || (b && c)` is more clear than `a || b && c` - the explicit parenthesization (especially when split over 3 lines with a different operator at the end of each line like this) improves clarity, I think. I read this like 3 times and wasn't sure if it was equivalent to `(a || b) && c` (which didn't make sense) at first, before I remember that `&&` was, in fact, higher presedence than `||` (and not equivalent).